mindmeld/components/nlp.py

```diff
@@ -1416,9 +1429,82 @@ def process_query(self, query, allowed_nlp_classes=None, dynamic_resource=None,
+            max_ngram_search (int): The max ngram value we want to break the input query
+
+        Returns:
+            list: A list of lists of non-overlapping entities for each n-best transcript
+        """
+        # This code block implements allowed entities described here:
+        # https://github.com/cisco/mindmeld/pull/280
+
+        # We extract the `entity` key from each token since that represents the normalized text of
+        # each token which we will compare against the normalized text of the query
+
+        def tokenize_to_tuples(text):
+            return tuple([token["entity"] for token in
+                          self.resource_loader.query_factory.tokenizer.tokenize(text)])
+
+        dynamic_gazetteer = dynamic_resource.get(GAZETTEER_RSC, {}) if dynamic_resource else {}
+        n_best_entities = [[] for _ in range(len(query))]
+
+        for entity in allowed_nlp_classes:
+            # check if entity is a system entity
+            if entity.startswith(SYSTEM_ENTITY_PREFIX):
+                for idx, n_best_query in enumerate(query):
+                    for sys_entity in n_best_query.system_entity_candidates:
+                        if sys_entity.entity.type == entity:
+                            n_best_entities[idx].append(sys_entity)
+                continue
+
+            # check if entity is in the gazetteers
+            # TODO: We are tokenizing every phrase in the gazetteer.
+            # This is a large CPU intensive op
+            consolidate_list = {**self.resource_loader.get_gazetteer(entity)['pop_dict'],
+                                **dynamic_gazetteer.get(entity, {})}
+            consolidated_set = {tokenize_to_tuples(key) for key in consolidate_list.keys()}
+
+            for idx, n_best_query in enumerate(query):
+                for ngram, token_span in get_ngrams_upto_n(n_best_query.normalized_tokens,
+                                                           max_ngram_search):
+                    if ngram not in consolidated_set:
+                        continue
+
+                    start_token = n_best_query.get_verbose_normalized_tokens()[token_span[0]]
+                    end_token = n_best_query.get_verbose_normalized_tokens()[token_span[1]]
+
+                    start_index = start_token['raw_start']
+                    end_index = end_token['raw_start'] + len(end_token['entity']) - 1
+                    span = Span(start_index, end_index)
+
+                    entity_val = Entity(
+                        text=n_best_query.text[start_index: end_index + 1],
+                        entity_type=entity
+                    )
+                    query_entity = QueryEntity.from_query(
+                        query=n_best_query, span=span, entity=entity_val
+                    )
+                    n_best_entities[idx].append(query_entity)
+
+        return [tuple(NestedEntity.get_largest_non_overlapping_entities(
```

Comment:
Open question: While I can't think of a good example for this, is there any use-case where we'd want to do this based on entity type?

Reply:
Good question. I think this is for the case were we remove an entity of the same type that overlaps with an entity of another type. I would presume the developer would have preprocessed entities of his/her interest before calling this function.